10_construct-database/medians.qmd

```diff
@@ -0,0 +1,117 @@
+---
+title: "Calculate Medians for Dashboard"
+author: "Aaron R. Williams"
+date: today
+format:
+  html:
+    toc: true
+    toc_float: true
+    embed-resources: true
+execute: 
+  warning: false
+editor_options: 
+  chunk_output_type: console
+---
+
+## Load Packages
+
+```{r}
+#| label: load-packages
+
+options(scipen = 999)
+
+library(tidyverse)
+library(here)
+library(tidycensus)
+
+```
+
+## Load Data
+
+```{r}
+#| label: load-data
+metrics <- read_csv(here("data", "00_mobility-metrics_recent.csv")) |>
+  select(-ends_with("_quality"), -ends_with("_lb"), -ends_with("_ub"))
+
+```
+
+We load county-level population estimates from the 2020 decennial census.
+
+```{r}
+population <- get_decennial(
+  geography = "county", 
+  variables = "P1_001N", 
+  year = 2020
+) |>
+  mutate(
+    state = str_sub(GEOID, start = 1, end = 2),
+    county = str_sub(GEOID, start = 3, end = 5)
+  ) |>
+  select(state, county, population = value)
+
+metrics <- left_join(metrics, population, by = c("state", "county"))
+
+sum(is.na(metrics$population))
+
+```
+
+## Calculate Medians
+
+We calculate weighted medians for each numeric variable. This means our estimates represent the county mobility metric for the average American. 
+
+We exclude missing values and calculate the proportion of counties and proportion of population missing for each metric. 
+
+```{r}
+missing_share <- function(x, weight) {
+  
+  sum(as.numeric(is.na(x)) * weight) / sum(weight)
+  
+}
+
+```
+
+Calculate 
+
+* The population-weighted median
+* The proportion of counties that are missing
+* The proportion of people that are missing
+
+```{r}
+weighted_median <- metrics |>
+  select(where(is.numeric)) |>
+  summarize(
+    across(
+      .cols = everything(), 
+      .fns = ~Hmisc::wtd.quantile(., weights = population, probs = 0.5, na.rm = TRUE)
+    )
+  ) |>
+  pivot_longer(cols = everything(), values_to = "weighted_median")
+
+missing_population <- metrics |>
+  select(where(is.numeric)) |>
+  summarize(
+    across(
+      .cols = everything(), 
+      .fns = ~missing_share(.x, weight = population)
+    )
+  ) |>
+  pivot_longer(cols = everything(), values_to = "missing_population")
+
+missing_county <- metrics |>
+  select(where(is.numeric)) |>
+  summarize(
+    across(
+      .cols = everything(), 
+      .fns = ~sum(is.na(.x)) / n()
+    )
+  ) |>
+  pivot_longer(cols = everything(), values_to = "missing_county")
+
+weighted_median |>
+  full_join(missing_population, by = "name") |>
+  full_join(missing_county, by = "name") |>
+  filter(name != "population") |>
+  print(n = Inf)
```

Comment:
I don't have a solution but should we consider a different calculation of median for metrics where there is such a high share of missing counties? I'm thinking in particular about the share_black_nh_exposure - I'm guessing there is a lot of missing here becuase the number of Black non-hispanic households is zero in many counties but that should be reflected in the median. My assumption on why there are NAs may be wrong.
   
name                                  weighted_median missing_population missing_county
   <chr>                                           <dbl>              <dbl>          <dbl>
 1 share_in_preschool                             0.388          0.000629         0.00191 
 2 rate_learning                                  0.984          0.0169           0.0331  
 3 meps20_black                                   0.507          0.000791         0.0312  
 4 meps20_hispanic                                0.493          0.000178         0.00668 
 5 meps20_white                                   0.157          0.0000397        0.00191 
 6 share_hs_degree                                0.915          0.0000296        0.000955
 7 share_digital_access                           0.891          0.00000125       0.000636
 8 pctl_income_20                             26341.             0.0000296        0.000955
 9 pctl_income_50                             60786.             0.0000296        0.000955
10 pctl_income_80                            121572.             0.0000296        0.000955
11 share_debt_col                                 0.259          0.000349         0.0270  
12 rate_low_birth_weight                          0.0811         0                0       
13 rate_injury_death                             72.9            0.000527         0.0340  
14 share_affordable_80_ami                        1.52           0.0000296        0.000955
15 share_affordable_50_ami                        1.37           0.0000296        0.000955
16 share_affordable_30_ami                        1.19           0.0000296        0.000955
17 count_homeless                              1400              0.00881          0.0945  
18 share_homeless                                 0.0214         0.00881          0.0945  
19 share_election_turnout                         0.681          0.00221          0.00986 
20 share_desc_rep_asian_other                     0.0790         0                0       
21 share_desc_rep_black_nonhispanic               0.0765         0                0       
22 share_desc_rep_hispanic                        0.120          0                0       
23 share_desc_rep_white_nonhispanic               0.611          0                0       
24 share_poverty_exposure                         0.0377         0.00000928       0.00223 
25 share_black_nh_exposure                        0.805          0.00840          0.144   
26 share_hispanic_exposure                        0.789          0.00124          0.0290  
27 share_other_nh_exposure                        0.880          0.000441         0.0207  
28 share_white_nh_exposure                        0.298          0.00000130       0.000636
29 transportation_cost                           56.0            0.0000293        0.000636
30 index_transit_trips                           54.4            0.0000293        0.000636
31 ratio_high_low_ses_fb_friends                  0.785          0.00176          0.0401  
32 count_membership_associations_per_10k          8.2            0.00188          0.0484  
33 rate_violent_crime                           894.             0.230            0.141   
34 rate_property_crime                         2356.             0.230            0.141   
35 rate_juv_arrest                              728.             0.235            0.146   
36 share_employed                                 0.808          0.0000296        0.000955
37 ratio_average_to_living_wage                   0.694          0.0000296        0.000955